microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java

```diff
@@ -0,0 +1,279 @@
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import java.io.File;
+import java.nio.ByteBuffer;
+
+
+public class AbstractSslEngineBenchmark extends AbstractMicrobenchmark {
+
+    private static final String PROTOCOL_TLS_V1_2 = "TLSv1.2";
+
+    public enum SslEngineProvider {
+        JDK {
+            @Override
+            SslProvider sslProvider() {
+                return SslProvider.JDK;
+            }
+        },
+        OPENSSL {
+            @Override
+            SslProvider sslProvider() {
+                return SslProvider.OPENSSL;
+            }
+        };
+        private final SslContext clientContext = newClientContext();
+        private final SslContext serverContext = newServerContext();
+
+        private SslContext newClientContext() {
+            try {
+                return SslContextBuilder.forClient()
+                        .sslProvider(sslProvider())
+                        .trustManager(InsecureTrustManagerFactory.INSTANCE)
+                        .build();
+            } catch (SSLException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        private SslContext newServerContext() {
+            try {
+                File keyFile = new File(getClass().getResource("test_unencrypted.pem").getFile());
+                File crtFile = new File(getClass().getResource("test.crt").getFile());
+
+                return SslContextBuilder.forServer(crtFile, keyFile)
+                        .sslProvider(sslProvider())
+                        .build();
+            } catch (Exception e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        SSLEngine newClientEngine(String cipher) {
+            return configureEngine(clientContext.newEngine(PooledByteBufAllocator.DEFAULT), cipher);
+        }
+
+        SSLEngine newServerEngine(String cipher) {
+            return configureEngine(serverContext.newEngine(PooledByteBufAllocator.DEFAULT), cipher);
+        }
+
+        abstract SslProvider sslProvider();
+
+        static SSLEngine configureEngine(SSLEngine engine, String cipher) {
+            engine.setEnabledProtocols(new String[]{ PROTOCOL_TLS_V1_2 });
+            engine.setEnabledCipherSuites(new String[]{ cipher });
+            return engine;
+        }
+    }
+
+    public enum BufferType {
+        HEAP {
+            @Override
+            ByteBuffer newBuffer(int size) {
+                return ByteBuffer.allocate(size);
+            }
+        },
+        DIRECT {
+            @Override
+            ByteBuffer newBuffer(int size) {
+                return ByteBuffer.allocateDirect(size);
+            }
+
+            @Override
+            void freeBuffer(ByteBuffer buffer) {
+                PlatformDependent.freeDirectBuffer(buffer);
+            }
+        };
+
+        abstract ByteBuffer newBuffer(int size);
+
+        void freeBuffer(ByteBuffer buffer) { }
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)